### PR TITLE
Ironic aio

### PR DESCRIPTION
**Pull request by derekhiggins**
Add a starting point for ironic on the bootstrap node, basically working but very much WIP.

### CODE REVIEW

File: tripleo-quickstart-config/roles/libvirt/setup/overcloud/templates/instackenv.json.j2 [hunk 1]
```diff
@@ -25,7 +25,7 @@
           "pm_type": "ipmi",
         {% endif %}
         "pm_user": "admin",
-        "pm_addr": "{{ undercloud_local_ip|default(undercloud_network_cidr)|nthhost(1) }}",
+        "pm_addr": "192.168.122.1",
```
Comment (hardys) [inline tripleo-quickstart-config/roles/libvirt/setup/overcloud/templates/instackenv.json.j2]: Hard-coding this seems fine for now, but there's also a bunch of logic related to this at the top of the template we can probably also remove - I can push a follow-up

File: 05_run_ocp.sh [hunk 2]
```diff
@@ -64,7 +64,7 @@ sudo cp ocp/bootstrap.ign ${IGN_FILE}
 ./get_rhcos_image.sh
 LATEST_IMAGE=$(ls -ltr redhat-coreos-maipo-*-qemu.qcow2 | tail -n1 | awk '{print $9}')
 sudo cp $LATEST_IMAGE /var/lib/libvirt/images/${CLUSTER_NAME}-bootstrap.qcow2
-virt-install --connect qemu:///system \
+sudo virt-install --connect qemu:///system \
```
Comment (hardys) [inline 05_run_ocp.sh]: Any idea why this is needed, when all we're doing is generating the xml?  Fine to add it but weird that I don't in my env (the user is in the libvirt group)

File: tripleo-quickstart-config/roles/libvirt/setup/overcloud/templates/instackenv.json.j2 (hunk 1 above)
Comment (derekhiggins) [reply tripleo-quickstart-config/roles/libvirt/setup/overcloud/templates/instackenv.json.j2]: Ack, Sounds good to me

File: 05_run_ocp.sh (hunk 2 above)
Comment (derekhiggins) [reply 05_run_ocp.sh]: I didn't look to hard to be honest, my stack user wasn't in the virt group so thats probably it

File: README.md [hunk 3]
```diff
@@ -45,6 +46,21 @@ e.g `sudo echo "192.168.122.235 ostest-api.test.metalkube.org" >> /etc/hosts`.
 Then you can interact with the k8s API on the bootstrap VM e.g
 `sudo oc status --verbose --config /etc/kubernetes/kubeconfig`.
 
+You can also use the ironic by talking to its API with the openstack client (from the
+virt host)
+`export OS_TOKEN=fake-token`
+`export OS_URL=http://ostest-api.test.metalkube.org:6385/`
+
+To Define a NODE (Get \<MAC\> and \<IPMI\_PORT\> from instackenv.json
+`NODE_UUID=$(openstack baremetal node create --driver ipmi --driver-info ipmi_address=192.168.122.1 --driver-info ipmi_port=6230 --driver-info ipmi_username=admin --driver-info ipmi_password=password --driver-info deploy_kernel=http://172.22.0.1/images/tinyipa-stable-rocky.vmlinuz --driver-info deploy_ramdisk=http://172.22.0.1/images/tinyipa-stable-rocky.gz | grep " uuid " | awk '{print $4}')`
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply README.md]: will do

File: README.md [hunk 4]
```diff
@@ -45,6 +46,21 @@ e.g `sudo echo "192.168.122.235 ostest-api.test.metalkube.org" >> /etc/hosts`.
...
+`openstack baremetal port create 00:32:49:d0:63:75 --node $NODE_UUID`
+`openstack baremetal node set $NODE_UUID --instance-info image_source=https://download.cirros-cloud.net/0.3.5/cirros-0.3.5-x86_64-disk.img --instance-info image_checksum=f8ab98ff5e73ebab884d80c9dc9c7290 --instance-info root_gb=10`
+`openstack baremetal node manage $NODE_UUID`
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply README.md]: willdo

File: ironic/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+RUN mkdir -p /var/www/html/images
+RUN curl http://tarballs.openstack.org/ironic-python-agent/tinyipa/tinyipa-stable-rocky.tar.gz | tar -C /var/www/html/images/ -xzf -
+
+RUN cp /etc/ironic/ironic.conf /etc/ironic/ironic.conf_orig
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT auth_strategy noauth
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT my_ip 172.22.0.1
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT debug true
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT default_network_interface noop
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT enabled_boot_interfaces pxe,ipxe
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT default_boot_interface ipxe
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply ironic/Dockerfile]: Doesn't look like it, I needed to specifically enable ipxe (I guess we enable it in devstack but the default should be changed?)
https://github.com/openstack/ironic/blob/5471157e4fdb5a735a29d88f5ad3af0d4b765ea3/ironic/conf/default.py#L105-L107

File: ironic/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+RUN crudini --set /etc/ironic/ironic.conf database connection sqlite:///ironic.db
+RUN crudini --set /etc/ironic/ironic.conf dhcp dhcp_provider none
+RUN crudini --set /etc/ironic/ironic.conf conductor automated_clean false
+RUN crudini --set /etc/ironic/ironic.conf conductor api_url http://172.22.0.1:6385
+RUN crudini --set /etc/ironic/ironic.conf deploy http_url http://172.22.0.1
+RUN crudini --set /etc/ironic/ironic.conf deploy http_root /var/www/html/
+RUN crudini --set /etc/ironic/ironic.conf pxe ipxe_enabled true
+RUN crudini --set /etc/ironic/ironic.conf pxe pxe_config_template /usr/lib/python2.7/site-packages/ironic/drivers/modules/ipxe_config.template
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply ironic/Dockerfile]: will do

File: ironic/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT default_network_interface noop
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT enabled_boot_interfaces pxe,ipxe
+RUN crudini --set /etc/ironic/ironic.conf DEFAULT default_boot_interface ipxe
+RUN crudini --set /etc/ironic/ironic.conf database connection sqlite:///ironic.db
+RUN crudini --set /etc/ironic/ironic.conf dhcp dhcp_provider none
+RUN crudini --set /etc/ironic/ironic.conf conductor automated_clean false
+RUN crudini --set /etc/ironic/ironic.conf conductor api_url http://172.22.0.1:6385
+RUN crudini --set /etc/ironic/ironic.conf deploy http_url http://172.22.0.1
+RUN crudini --set /etc/ironic/ironic.conf deploy http_root /var/www/html/
+RUN crudini --set /etc/ironic/ironic.conf pxe ipxe_enabled true
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply ironic/Dockerfile]: I seemed to need it, I'll try again without it now that I have was working example

File: README.md [hunk 8]
```diff
@@ -45,6 +46,21 @@ e.g `sudo echo "192.168.122.235 ostest-api.test.metalkube.org" >> /etc/hosts`.
...
+`openstack baremetal node provide $NODE_UUID`
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply README.md]: will do

File: README.md [hunk 9]
```diff
@@ -45,6 +46,21 @@ e.g `sudo echo "192.168.122.235 ostest-api.test.metalkube.org" >> /etc/hosts`.
 Then you can interact with the k8s API on the bootstrap VM e.g
 `sudo oc status --verbose --config /etc/kubernetes/kubeconfig`.
 
+You can also use the ironic by talking to its API with the openstack client (from the
+virt host)
+`export OS_TOKEN=fake-token`
+`export OS_URL=http://ostest-api.test.metalkube.org:6385/`
```
Comment (hardys) [inline README.md]: We need to either add ostest-api to /etc/hosts in the scripts or mention above that you need to add it based on the libvirt leases output.

File: ironic/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply ironic/Dockerfile]: Oh, right. You need enabled_boot_interfaces, but probably not default_boot_interface.

File: ironic/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply ironic/Dockerfile]: I think changing enabled_boot_interfaces is enough for new nodes to catch the ipxe boot interface.

File: README.md (hunk 9 above)
Comment (derekhiggins) [reply README.md]: We already mention it above,
"""You can then add the IP to the /etc/hosts on the node with the hostname,
e.g `sudo echo "192.168.122.235 ostest-api.test.metalkube.org" >> /etc/hosts`."""

File: ironic/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (derekhiggins) [reply ironic/Dockerfile]: looks like its needed, when I remove it I get
2019-01-31 11:39:48.363 9 ERROR ironic.conductor.utils [req-6f694a70-697a-49bf-9ea4-0fda711970e2 - - - - -] Error while preparing to deploy to node 858d95d4-04b4-4b59-bb94-7dcb89c0407d: Could not link image http://172.22.0.1/images/tinyipa-stable-rocky.gz from /tftpboot/master_images/ab9da2a6-341e-5684-8705-6d1f6a451a97.converted to /tftpboot/858d95d4-04b4-4b59-bb94-7dcb89c0407d/deploy_ramdisk, error: [Errno 2] No such file or directory: ImageDownloadFailed: Could not link image http://172.22.0.1/images/tinyipa-stable-rocky.gz from /tftpboot/master_images/ab9da2a6-341e-5684-8705-6d1f6a451a97.converted to /tftpboot/858d95d4-04b4-4b59-bb94-7dcb89c0407d/deploy_ramdisk, error: [Errno 2] No such file or directory

File: ironic/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+RUN mkdir /tftpboot
+
+RUN ironic-dbsync --config-file /etc/ironic/ironic.conf create_schema
+
+RUN echo "#!/usr/bin/bash" >> /bin/runironic
+RUN echo "/usr/lib/rabbitmq/bin/rabbitmq-server > /var/log/rabbit.out 2>&1 &" >> /bin/runironic
+RUN echo "/usr/bin/python2 /usr/bin/ironic-conductor > /var/log/ironic-conductor.out 2>&1 &" >> /bin/runironic
+RUN echo "/usr/bin/python2 /usr/bin/ironic-api > /var/log/ironic-api.out 2>&1 &" >> /bin/runironic
+RUN echo "/usr/sbin/httpd" >> /bin/runironic
+RUN echo "/usr/sbin/dnsmasq -d -q -i eth1 --dhcp-range 172.22.0.10,172.22.0.100 --enable-tftp --tftp-root=/tftpboot --dhcp-boot=http://172.22.0.1/boot.ipxe > /var/log/dnsmasq.out 2>&1 &" >> /bin/runironic
```
[the remark this answers is not in the document]
Comment (derekhiggins) [reply ironic/Dockerfile]: I'll take a look at adding one if it looks like this AIO approach is going to stick around.

File: ironic/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply ironic/Dockerfile]: This should not be happening.. I wonder if we have an issue in ironic.

**Pull request merged**